### PR TITLE
add tests for edge cases for all operators

### PR DESCRIPTION
**Pull request by arvidn**
please review the test cases and the expected output. I did and discovered a few missing checks that I'm adding in this patch.

### CODE REVIEW

File: tests/edge-cases/point_add-05.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+brun '(point_add (q (1 2)) (q (1 2)))'
+FAIL: point_add expects blob, got ff01ff0280: from_bytes(): incompatible function arguments. The following argument types are supported:
+    1. (self: buffer) -> blspy.G1Element
+
+Invoked with: None ((q 2) (q 2))
```
Comment (arvidn) [inline tests/edge-cases/point_add-05.txt:5]: I'm not sure about this one. Is this correct? the output doesn't follow the pattern used by other operator failures, which I've learned sometimes indicate the wrong kind of exception is thrown (i.e. an unexpected exception)

File: tests/edge-cases/apply-02.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+brun '(a (q 1))'
+FAIL: unimplemented operator 3
```
Comment (arvidn) [inline tests/edge-cases/apply-02.txt:2]: this is a bit questionable as well, but it seems odd to have `a` in the operator list at all, before it's implemented

File: clvm/more_ops.py [hunk 3]
```diff
@@ -152,15 +152,19 @@ def op_point_add(items):
 
 
 def op_strlen(args):
+    if args.list_len() != 1:
+        raise EvalError("strlen takes exactly 1 argument", args)
     a0 = args.first()
     if a0.pair:
-        raise EvalError("len on list", a0)
+        raise EvalError("strlen on list", a0)
```
Comment (arvidn) [inline clvm/more_ops.py:159]: this affects one of the existing test cases. I think `strlen` is better than `len` here.

File: tests/edge-cases/if-07.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+brun '(i (q (1 2)) (q (1 2)) (q (1 2)))'
+(q 2)
```
Comment (arvidn) [inline tests/edge-cases/if-07.txt:2]: Is i intentional to treat a (non-empty) list as `true`?

File: tests/edge-cases/logand-01.txt [hunk 5]
```diff
@@ -0,0 +1,2 @@
+brun '(logand)'
+-1
```
Comment (arvidn) [inline tests/edge-cases/logand-01.txt:2]: @aqk this behavior is documented, right?

File: tests/edge-cases/lognot-04.txt [hunk 6]
```diff
@@ -0,0 +1,2 @@
+brun '(lognot ())'
+-1
```
Comment (arvidn) [inline tests/edge-cases/lognot-04.txt:2]: @aqk and this behavior, it is not necessarily obvious

File: tests/edge-cases/point_add-06.txt [hunk 7]
```diff
@@ -0,0 +1,2 @@
+brun '(point_add (q 0xffff) (q 0xffff))'
+FAIL: point_add expects blob, got 82ffff: Length of bytes object not equal to G1Element::SIZE (0xffff 0xffff)
```
Comment (arvidn) [inline tests/edge-cases/point_add-06.txt:2]: it's a little bit odd that the error message prints the *serialized* form of the value it was passed

File: clvm/core_ops.py [hunk 8]
```diff
@@ -47,6 +47,8 @@ def op_raise(args):
 
 
 def op_eq(args):
+    if args.list_len() != 2:
```
Comment (matt-o-how) [inline clvm/core_ops.py:50]: I thought that we wanted equal to take unlimited arguments so you could assert equality between multiple things without needing multiple checks?

Comment (arvidn) [reply clvm/core_ops.py:50]: perhaps. that's not what it does now though. If we make that change, we can change this check too.

Comment (arvidn) [reply clvm/core_ops.py:50]: I'm not sold on that being desirable though, at least not in the low-level language. There's consistency with `>` and `>s` right now (in being binary operators).

**Comment by arvidn** [comment]
@aqk @richardkiss ping

File: tests/edge-cases/if-07.txt (hunk 4 above)
Comment (richardkiss) [reply tests/edge-cases/if-07.txt:2]: Yes

File: tests/edge-cases/point_add-05.txt (hunk 1 above)
Comment (richardkiss) [reply tests/edge-cases/point_add-05.txt:5]: Yeah, probably worth checking the pair/atom type and raising a different message prior to trying to evaluate.

File: tests/edge-cases/point_add-06.txt (hunk 7 above)
Comment (richardkiss) [reply tests/edge-cases/point_add-06.txt:2]: Hmm, that is odd. Probably could do with a realignment.

File: tests/edge-cases/softfork-02.txt [hunk 9]
```diff
@@ -0,0 +1,2 @@
+brun '(softfork (q 1))'
+()
```
Comment (richardkiss) [inline tests/edge-cases/softfork-02.txt:2]: Maybe the `softfork` tests should include `-c` so costs are output.

File: tests/edge-cases/point_add-06.txt (hunk 7 above)
Comment (arvidn) [reply tests/edge-cases/point_add-06.txt:2]: when an `SExp` is printed (via `__str__`) it uses `to_bin()` which in turn uses the serializer. I think it would be better to keep the serialization separate. Ideally we could change the serialization format without changing the way an `SExp` is printed. I will propose a patch for this

Comment (arvidn) [reply tests/edge-cases/point_add-06.txt]: I've made this change, and removed `to_bin()`. Please take a look

File: tests/edge-cases/point_add-05.txt (hunk 1 above)
Comment (arvidn) [reply tests/edge-cases/point_add-05.txt]: done, please review

**Comment by arvidn** [comment]
I've updated this PR with additional commits to make it easier to review. I intend to squash once it's in good shape

File: clvm/SExp.py [hunk 10]
```diff
@@ -170,7 +164,7 @@ def as_python(self):
         return as_python(self)
 
     def __str__(self):
-        return self.as_bin().hex()
+        return self.atom.hex()
```
Comment (arvidn) [inline clvm/SExp.py:167]: this won't work for cons objects

Comment (richardkiss) [reply clvm/SExp.py:167]: Yes, seems like this will change test outputs, as `as_bin` did serialization.

File: clvm/core_ops.py (hunk 8 above)
Comment (richardkiss) [reply clvm/core_ops.py:50]: Maybe I'm paranoid, but I am also afraid that `=` with multiple values could lead to subtle consensus bugs across implementations.

File: clvm/SExp.py (hunk 10 above)
Comment (arvidn) [reply clvm/SExp.py:167]: the intention was to change the test output, or rather, the error messages. But when we actually serialize things with the serializer, we don't use `as_bin()`, we use the serializer. All tests still pass with this change, it's just there is no test for this `__str__` function yet, and it would fail for non-atoms.

File: tests/edge-cases/divmod-01.txt [hunk 11]
```diff
@@ -0,0 +1,2 @@
+brun '(divmod)'
```
Comment (aqk) [inline tests/edge-cases/divmod-01.txt:1]: @arvidn @richardkiss If we only return one value, we should consider changing the name `divmod` -> e.g. `div`

File: tests/edge-cases/divmod-09.txt [hunk 12]
```diff
@@ -0,0 +1,2 @@
+brun '(divmod (q -1) (q -1))'
```
Comment (aqk) [inline tests/edge-cases/divmod-09.txt:1]: We should check all sign combinations. One of those weird int math corners.

https://www.evanjones.ca/integer-division.html

File: tests/edge-cases/list-04.txt [hunk 13]
```diff
@@ -0,0 +1,2 @@
+brun '(l ())'
```
Comment (aqk) [inline tests/edge-cases/list-04.txt:1]: This is a good one. I mention this in the doc, but I'll try to mention it several times.

File: tests/edge-cases/logand-01.txt (hunk 5 above)
Comment (aqk) [reply tests/edge-cases/logand-01.txt:2]: Yes, it's in the table of outputs with zero arguments, and in the arithmetic unity table, but there it's described as "all ones", so I'll improve that.

File: tests/edge-cases/logand-05.txt [hunk 14]
```diff
@@ -0,0 +1,2 @@
+brun '(logand (q (1 2)) (q (1 2)))'
+FAIL: logand requires int args ((q 2) (q 2))
```
Comment (aqk) [inline tests/edge-cases/logand-05.txt:2]: This error message would be hard to understand if I hadn't spent so much time with clvm. I know this is not new in your patch.

File: tests/edge-cases/logand-07.txt [hunk 15]
```diff
@@ -0,0 +1,2 @@
+brun '(logand (q 128) (q 128))'
```
Comment (aqk) [inline tests/edge-cases/logand-07.txt:1]: For the logical operators, I feel like it would be good to include a value that spans one extra byte, like 0x01ff

File: tests/edge-cases/lognot-02.txt [hunk 16]
```diff
@@ -0,0 +1,2 @@
+brun '(lognot (q 1))'
+-2
```
Comment (aqk) [inline tests/edge-cases/lognot-02.txt:2]: Making a note to explain this explicitly & well in the docs.

File: tests/edge-cases/lognot-06.txt [hunk 17]
```diff
@@ -0,0 +1,2 @@
+brun '(lognot (q 0xffff))'
```
Comment (aqk) [inline tests/edge-cases/lognot-06.txt:1]: Nice one.

File: tests/edge-cases/lsh-01.txt [hunk 18]
```diff
@@ -0,0 +1,2 @@
+brun '(lsh)'
+FAIL: ash requires 2 args ()
```
Comment (aqk) [inline tests/edge-cases/lsh-01.txt:2]: Hold the phone. opcode is `lsh`, but message says `ash`

File: tests/edge-cases/not-05.txt [hunk 19]
```diff
@@ -0,0 +1,2 @@
+brun '(not (q (1 2)))'
```
Comment (aqk) [inline tests/edge-cases/not-05.txt:1]: `brun '(not (q (1 2)))'`
@richardkiss This seems like it should be an error to me.

File: clvm/more_ops.py (hunk 3 above)
Comment (aqk) [reply clvm/more_ops.py:161]: I agree. `strlen` is more direct.

File: tests/edge-cases/all-01.txt [hunk 20]
```diff
@@ -0,0 +1,2 @@
+brun '(all)'
+1
```
[the remark this answers is not in the document]
Comment (matt-o-how) [reply tests/edge-cases/all-01.txt:2]: Why not enforce that it has at least 1 parameter? I can't think of a circumstance where you would want to call it without any parameters.

File: clvm/core_ops.py (hunk 8 above)
Comment (aqk) [reply clvm/core_ops.py:50]: I agree, better to be conservative and only allow two arguments at the lowest level.

File: tests/edge-cases/all-01.txt (hunk 20 above)
[the remark this answers is not in the document]
Comment (aqk) [reply tests/edge-cases/all-01.txt:2]: Yes, that's what I'm saying too. It's not useful with zero args.

File: tests/edge-cases/apply-02.txt (hunk 2 above)
Comment (aqk) [reply tests/edge-cases/apply-02.txt:2]: No problem. This documents that it doesn't work in this branch.
I think that's good.

File: tests/edge-cases/all-01.txt (hunk 20 above)
[the remark this answers is not in the document]
Comment (richardkiss) [reply tests/edge-cases/all-01.txt:2]: Here's one I can think of: when used with the new `apply`, you might do `(a #all ITEMS)`; it would save you checking against ITEMS being zero. Note that in python, `any` and `all` each accept zero parameters.

File: tests/edge-cases/ash-02.txt [hunk 21]
```diff
@@ -0,0 +1,2 @@
+brun '(ash (q 1))'
+FAIL: ash requires 2 args (q)
```
[the remark this answers is not in the document]
Comment (richardkiss) [reply tests/edge-cases/ash-02.txt:2]: One option is the s-expression in the error message can be decompiled without keywords (since it generally just includes already-evaluated expressions).

File: tests/edge-cases/divmod-01.txt (hunk 11 above)
Comment (richardkiss) [reply tests/edge-cases/divmod-01.txt:1]: This hasn't been changed yet, as we haven't come to a consensus on `div` vs `divmod`. Let me run a quick test.

Comment (richardkiss) [reply tests/edge-cases/divmod-01.txt:1]:
```
kiss@oasis ~ % python -m timeit 'A = 74738383914845; B = 83717484' "divmod(A,B)"
10000000 loops, best of 3: 0.164 usec per loop
kiss@oasis ~ % python -m timeit 'A = 74738383914845; B = 83717484' "A/B"        
10000000 loops, best of 3: 0.0503 usec per loop
kiss@oasis ~ % python -m timeit 'A = 74738383914845; B = 83717484' "A%B"
10000000 loops, best of 3: 0.0497 usec per loop
```

Okay, this is pretty wack. I don't know what to make of these results.

Comment (richardkiss) [reply tests/edge-cases/divmod-01.txt:1]: It's almost as if it's spending most of its time making the tuple.

Comment (richardkiss) [reply tests/edge-cases/divmod-01.txt:1]: Okay, let's use integer division.
```
kiss@oasis ~ % python -m timeit 'A = 74738383914845; B = 83717484' "A//B"       
10000000 loops, best of 3: 0.0522 usec per loop
kiss@oasis ~ % python -m timeit 'A = 74738383914845; B = 83717484' "A%B" 
10000000 loops, best of 3: 0.0512 usec per loop
kiss@oasis ~ % python -m timeit 'A = 74738383914845; B = 83717484' "(A//B, A%B)"
10000000 loops, best of 3: 0.117 usec per loop
kiss@oasis ~ % python -m timeit 'A = 74738383914845; B = 83717484' "divmod(A,B)"
10000000 loops, best of 3: 0.166 usec per loop
kiss@oasis ~ % python -m timeit 'A = 74738383914845; B = 83717484' "Q = A//B; (Q, A - B * Q)"
10000000 loops, best of 3: 0.104 usec per loop
```

Seems like calculating the remainder using arithmetic is the fastest. I recommend we add integer division and keep divmod. Your thoughts @aqk @arvidn @bramcohen ?

File: tests/edge-cases/logand-05.txt (hunk 14 above)
Comment (richardkiss) [reply tests/edge-cases/logand-05.txt:2]: Maybe we should print just the first invalid argument instead of all of them.

File: tests/edge-cases/not-05.txt (hunk 19 above)
Comment (richardkiss) [reply tests/edge-cases/not-05.txt:1]: This is a boolean operator, so it ought to have the same semantics as `i`. This is how `i` works too.

File: tests/edge-cases/all-01.txt (hunk 20 above)
[the remark this answers is not in the document]
Comment (aqk) [reply tests/edge-cases/all-01.txt:2]: Yes, makes sense if you are generating params with other code.

File: tests/edge-cases/divmod-01.txt (hunk 11 above)
Comment (aqk) [reply tests/edge-cases/divmod-01.txt:1]: I'm ok with keeping both, as long as div(A,B) always means exactly divmod(A,B)[0]

File: clvm/SExp.py (hunk 10 above)
Comment (arvidn) [reply clvm/SExp.py:167]: I'm backing all of this out. I will propose a less intrusive alternative as a separate PR

File: tests/edge-cases/concat-04.txt [hunk 22]
```diff
@@ -0,0 +1,2 @@
+brun '(concat () ())'
+()
```
[the remark this answers is not in the document]
Comment (arvidn) [reply tests/edge-cases/concat-04.txt:2]: added as `concat-09.txt` and `concat-10.txt`

File: tests/edge-cases/concat-05.txt [hunk 23]
```diff
@@ -0,0 +1,2 @@
+brun '(concat (q (1 2)) (q (1 2)))'
+FAIL: concat on list (q 2)
```
[the remark this answers is not in the document]
Comment (arvidn) [reply tests/edge-cases/concat-05.txt:2]: added as `concat-11.txt` and `concat-12.txt`

**Comment by arvidn** [comment]
I've rebased this on top of develop. please make sure the changes in test cases are correct!

https://github.com/Chia-Network/clvm/pull/17/commits/0f5673466ca5960ba64d739ffa5481d7f2028309

I dropped the changes to `as_bin()` and `__str__`.

I believe the remaining comments are about `divmod` and `div`, which should be a separate PR. It's risky to rebase this since the existing (and reviewed) test cases may get mixed up in new changes that didn't have tests when they landed, so I think we should be careful to not widen the scope of this PR.

**Comment by arvidn** [comment]
I still think it makes sense to squash this after review, but I don't have strong feelings about it if it may delay landing it
